### PR TITLE
feat(native-chat): local attachments in the composer on Remote Orca Server panes

### LINKED ISSUE #6889: [Feature]: Support pasting local screenshots into SSH remote sessions
## Feature request

Support pasting or sending local clipboard images/screenshots into SSH-backed remote workspaces and agent sessions.

## Problem

When working in an Orca SSH remote workspace, the active agent/session runs on the remote host, but screenshots and copied images usually live on the local desktop clipboard.

Today this makes common workflows awkward. For example:

1. Connect to a remote SSH host in Orca.
2. Run an agent/session in that remote workspace.
3. Take a screenshot locally, or copy an image from the local browser/desktop.
4. Try to send that image to the remote agent.

There does not appear to be an easy Cursor/VS Code-like path that bridges the local clipboard image into the remote workspace/session. Users have to manually save the image locally, transfer it to the remote machine, then reference the remote path in the agent prompt.

## Desired behavior

From a remote SSH workspace/session, Orca should let users paste or attach local images/screenshots without manual file transfer.

Possible UX options:

- Paste an image directly into the remote agent prompt/chat and have Orca upload it to a temporary remote file, then attach/reference that file for the agent.
- Add a “Paste image from local clipboard” action for SSH workspaces.
- When an image is pasted into a remote terminal/session prompt, upload it through the existing SSH/SFTP/relay path and insert the remote path.
- Optionally show a short toast/status like “Uploaded screenshot to remote workspace”.

## Why this matters

Image and screenshot context is common in AI coding workflows:

- UI bug reports
- visual diffs
- error dialogs
- browser screenshots
- design references
- mobile/emulator screenshots

For local workspaces, sending image context is relatively straightforward. For SSH remotes, the local/remote boundary makes it much less convenient even though the user is still interacting with Orca locally.

## Suggested implementation direction

If feasible, reuse the existing SSH filesystem/SFTP/relay upload path rather than introducing a new transport:

1. Detect pasted clipboard image on the local desktop side.
2. Save it as a temporary file under a remote-safe location, for example inside the remote workspace or `~/.orca-remote/uploads/`.
3. Insert/attach the resulting remote path into the target agent/session.
4. Clean up old uploaded clipboard images eventually, or make the retention policy explicit.

## Edge cases / questions

- Should uploads be scoped to the workspace, the SSH target, or the session?
- What should happen when the SSH target is disconnected?
- Should filenames include timestamps and safe extensions (`.png`, `.jpg`, etc.)?
- Should large images be size-limited or compressed before upload?
- Should the user see the remote path that was created?
- Can this share behavior with any existing local image-paste support?

## Security / privacy considerations

- Do not upload clipboard images unless the user explicitly pastes/chooses the action.
- Avoid logging image contents or remote upload paths unnecessarily.
- Use safe generated filenames and avoid trusting clipboard-provided names.
- Make it clear the image is being copied from the local machine to the remote host.

### LINKED ISSUE #12572: [Feature]: Attach files by drag-and-drop and paste clipboard images into the chat composer on Orca Remote Server panes
### Problem or use case

Setup: Orca desktop 1.4.168 on Windows, paired to a headless `orca serve` runtime (v1.4.162) on a Linux VPS. Agent sessions run on the VPS; my files and screenshots are on the Windows machine.

In a runtime-owned (`remote:`) chat pane I cannot get local content to the agent:

- Dragging a file from Windows Explorer onto the chat composer does not attach it.
- Pasting a screenshot from the clipboard into the composer does nothing.

The only path is landing the file on the server out of band (scp, synced folder) and then typing the server-side path into the prompt. In practice that friction means the screenshot simply does not get sent and the agent works from a typed description instead. The cost is information loss, not keystrokes — I hit exactly this while diagnosing a rendering bug, where a re-typed fragment silently turned a newline into a space and sent the diagnosis down the wrong path for a while.

From the source this looks like a known, deliberate gap rather than a bug. `resolveNativeChatAttachmentOwner` in `src/renderer/src/components/native-chat/native-chat-attachment-upload.ts` returns `{ kind: 'runtime' }` for runtime-owned panes, with the comment:

> Runtime-owned (`remote:`) panes keep the composer's existing local-attachment block; runtime upload support is a separate seam.

The neighbouring transports are already handled: local panes attach directly, and SSH panes save the clipboard image on the remote host over SFTP (`saveClipboardImageAsTempFile({ connectionId })`) so the attached path names a file the remote agent can actually read.

### Proposed solution

Implement that runtime upload seam for the desktop composer, so all three attach surfaces work on runtime-owned panes:

- file drag-and-drop onto the composer,
- the file picker,
- clipboard image paste.

In each case: upload the bytes to the runtime over the existing pairing channel, write them to a temp location inside the workspace, and attach or insert the runtime-side path — the same end state the SSH path already produces.

The transport exists in at least one client already: Orca Mobile uploads images into a runtime-owned native chat session today (`mobile/src/session/mobile-native-chat-image-attachment.ts`, `use-mobile-native-chat-image-attachments.ts`). It is the desktop renderer that blocks.

Worth settling as part of this:

- per-file and aggregate size limits for client-originated bytes,
- what happens when the pairing connection drops mid-upload — a visible failure, not a silent no-op,
- retention and cleanup of uploaded temp files,
- whether arbitrary file types are inserted as `@path` references while images become attachment chips, matching local-pane behaviour.

### Alternatives or additional context

Current workaround: paste the image into a GitHub comment box without submitting it — GitHub uploads on paste and hands back a URL — then give the agent that URL to fetch. It works from any machine and needs no file transfer, but it routes private screenshots through GitHub, so it is not usable for anything sensitive.

Related:

- #6889 (P1, open) — pasting local screenshots into **SSH** remote sessions. I commented there asking for remote-server scope; filing this separately because it is a different transport seam and because it also covers arbitrary file drag-and-drop, not just clipboard images. Happy for it to be closed as a duplicate if you would rather track both under #6889.
- #9031 — file transfers in the paired **web Explorer** (Explorer, not the chat composer).
- #7777 — pasting an OS-file-manager-copied file into the terminal.

### LINKED ISSUE #17410: [Feature] Local attachments are not available for remote sessions.
<kbd>[cmer](https://discord.com/channels/1492228674081263786/1543627287411433563/1543627287411433563)</kbd> [cmer](https://discord.com/channels/1492228674081263786/1543627287411433563/1543627287411433563)  `via Discord`

I'm trying to set up a remote development environment and the fact that I can't share screenshots and other files with my remote agent is a definitely a blocker for me. Is there a way around it or is it possible to implement this feature?

### PR DESCRIPTION
**Pull request by cmer**
Fixes #12572
Fixes #17410
Fixes #6889

> **Stacked on #17787** — the first two commits here belong to that PR (the nested clipboard routing fix this feature builds on). Review the last commit; this PR will shrink to it once #17787 merges.

## Problem

On a runtime-owned (`remote:`) native chat pane, none of the composer's attach surfaces work: drag-and-drop and the file picker show *"Local attachments are not available for remote sessions"*, and clipboard image paste is refused. Meanwhile the same pane's **terminal** view happily uploads a pasted image to the server, and mobile clients attach images to the very same sessions — the gap was specific to the desktop/web composer. Users fall back to scp/synced folders or pasting screenshots into GitHub comment boxes for the URL (#12572, #6889 comments).

## What this does

All three composer attach surfaces now work on runtime-owned panes, producing the same end state the SSH composer path already produces — and it's wiring, not new transport:

- **Clipboard image paste** calls the same `saveClipboardImageAsTempFile({ runtimeEnvironmentId, connectionId })` terminal paste uses (chunked upload over the pairing channel, single-frame fallback for old servers). Nested worktrees on a server-owned SSH target work from day one (#17787).
- **Drag-and-drop and the file picker** upload through the existing `importExternalPathsToRuntime` into the worktree's `.orca/drops` (no-clobber), mirroring terminal drops — including the per-file skip/failure toasts and the capability assertion that surfaces *"Remote file changes require a newer Orca server"* against old servers.
- **The attachment owner resolver** now returns a full runtime owner (environment id, worktree, server-owned connection id, host expectation), failing closed to `not-ready` while ownership is mid-flip. A stale-owner guard fails an in-flight upload if the pane's host changes underneath it.
- **The attach funnel's invariant changed, not disappeared**: instead of blocking every remote pane, `attachResolvedPaths` receives owner-resolved host contexts. Client-local paths still can't reach a remote agent because every entry point resolves the owner and uploads first (the #6648 property, now enforced at the entries and pinned by tests).
- **Attachment chips carry their owning host**: image previews read via the runtime for in-worktree files (dropped images get real thumbnails); pasted server-`/tmp` images fall back to the generic thumbnail, matching current SSH-paste behavior. A runtime chip never carries a connection id — that connection is server-owned and must not be dialed from the client.

No new wire methods, no capability negotiation needed. The send path is untouched (runtime image sends already existed for mobile). The paired web client shares the renderer, so it gets all three surfaces too; mobile already worked and is unchanged.

## Testing

At the pre-agreed seams:

- `native-chat-attachment-upload.test.ts` — resolver returns the enriched runtime owner (plain, nested-SSH, focused-host, not-ready), and the runtime upload flow: import args, `.orca/drops` destination, skip/failure toasts, stale-owner failure, error → null.
- `use-native-chat-composer-paste.test.tsx` — runtime paste routes to the runtime save (plain + nested) and attaches the returned path; existing SSH/local/failure cases unchanged.
- `use-native-chat-external-attachments.test.tsx` — runtime drop/picker uploads and attaches destination-side paths with the runtime host context; failed uploads attach nothing; SSH/local/not-ready/disabled cases unchanged.

Full `pnpm test`, `pnpm tc`, oxlint, and the localization verifies pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019G22teMcYiNe9RmGQ4dKuK

### CODE REVIEW

**Comment by cmer** [comment]
Addressed Greptile's P1 (clipboard paste retains stale owner) in f5aad0f41:

- **The gap was real**: drop/picker uploads were fenced by `assertCurrent`, but clipboard paste attached the saved temp path without re-validating the owner after the async save — a pane re-pair, environment switch, or SSH generation bump mid-save could attach a path minted on the old host.
- **Fix**: after the save resolves, both paste flows (`handlePaste` and the menu-driven `pasteFromClipboard`) re-resolve the owner and compare host identity via a new shared `nativeChatAttachmentOwnersMatch` helper — the same comparison the drop/picker fence now uses (`assertNativeChatRuntimeOwnerCurrent` was refactored onto it). On drift the attachment is dropped with a visible notice ("Attachment upload host changed; retry the attach."), never silently. `not-ready` deliberately never matches: an unverifiable host is a changed host.
- **Tests**: field-level match semantics pinned in `native-chat-attachment-upload.test.ts` (identity match per owner kind, drift rejection incl. generation bumps and not-ready); the hook-level fence pinned in `use-native-chat-composer-paste.test.tsx` (owner changes mid-save → no attach + notice; unchanged owner still attaches).

Full native-chat suite (952 tests), typecheck, and lint pass.

File: src/renderer/src/components/native-chat/native-chat-attachment-upload.ts [hunk 1]
```diff
@@ -139,3 +184,41 @@ export async function uploadNativeChatAttachmentPaths(
     toast.dismiss(pending)
   }
 }
+
+function assertNativeChatRuntimeOwnerCurrent(owner: NativeChatRuntimeAttachmentOwner): void {
+  const current = resolveNativeChatAttachmentOwnerForWorktree(
```
[the remark this answers is not in the document]
Comment (cmer) [reply src/renderer/src/components/native-chat/native-chat-attachment-upload.ts:234]: Confirmed and fixed in 48ece5fee. The asymmetry was real: the tab-path capture resolves the worktree path with the `worktreesByRepo` fallback, while the assert's tab-less re-resolution used only `getKnownWorktreeById` — so during index hydration the assert could resolve `not-ready` and spuriously fail a legitimate attach. The structured branch (which the assert re-resolves through) now uses the same fallback, pinned by a test that captures through an empty known-worktree index.

**Comment by cmer** [comment]
Pullfrog's review is in (no critical issues, one inline edge-case) — addressed in 48ece5fee:

- **Stale-owner assert vs. hydration window**: the inline finding was correct — owner capture through the terminal-tab path resolves the worktree path with a `worktreesByRepo` catalog fallback, but the stale-owner assert re-resolved without the tab id and used only the known-worktree index, so a mid-upload assert during index hydration could fail a legitimate attach with "Attachment upload host changed." Both resolution branches now share the same fallback, so capture and re-resolution agree regardless of provenance. Pinned by a new resolver test (owner resolves via the catalog fallback with an empty known-worktree index).

Full native-chat suite (953 tests), typecheck, and lint pass. That closes out all bot findings on this PR: Greptile's P1 (paste fence, f5aad0f41) and pullfrog's inline edge-case (48ece5fee).
